### ISSUE TITLE
MySQLdb + db_pool trigger timeout exception and hang since 0.15.1 through 0.16.1

### ISSUE THREAD

**Issue by avillacis** [eventlet/eventlet#193]
I wrote a moderately complex program (Elastix endpointconfig2) using eventlet. This program launches parallel network probes using eventlet and writes the result to a MySQL database. The program works correctly on a CentOS 5 setup (python-2.4.3-56.el5 python-greenlet-0.3.1-2.el5 MySQL-python-1.2.3-0.1.c1.el5 python-eventlet-0.9.12-2.el5 using poll hub) and on a Fedora 18 ARM setup (python-2.7.3-13.fc18.armv5tel python-greenlet-0.3.1-4.fc15.armv5tel MySQL-python-1.2.3-6.fc18.armv5tel python-eventlet-0.9.17-2.fc18.noarch using epoll hub).

The program fails on a CentOS 7 setup (python-2.7.5-16.el7.x86_64 python-greenlet-0.4.2-3.el7.x86_64 MySQL-python-1.2.3-11.el7.x86_64 0.15.2-1.el7.noarch using epoll hub) with the following exception trace:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 346, in fire_timers
    timer()
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/timer.py", line 58, in __call__
    cb(*args, **kw)
  File "/usr/lib/python2.7/site-packages/eventlet/tpool.py", line 55, in tpool_trampoline
    _c = _rsock.recv(1)
  File "/usr/lib/python2.7/site-packages/eventlet/greenio.py", line 262, in recv
    timeout_exc=socket.timeout("timed out"))
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 155, in trampoline
    return hub.switch()
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 187, in switch
    return self.greenlet.switch()
timeout: timed out
```
By doing several reinstalls of python-eventlet, I have seen that the problem is in python-eventlet between 0.15 and 0.15.1.

git bisects points at the following commit:
```
f37a87b1f8e2f09b8a30c2e947486cf0858650cd is the first bad commit
commit f37a87b1f8e2f09b8a30c2e947486cf0858650cd
Author: Sergey Shepelev <temotor@gmail.com>
Date:   Thu Apr 24 17:42:30 2014 +0400

    greenio, tpool: python3 compatibility
    
    Also:
    - PEP-8
    - check both EAGAIN/EWOULDBLOCK
    - use system implementation of GreenPipe.readinto()

:040000 040000 529a31864c836fbe655cf4c1680de4b9f0752f4c 81d257d99807fe7527eb806d93a412382f867d0b M	eventlet
:040000 040000 311a5516d9e601b3717a337c0cd032bb96c2764a 138acaf6f68ef4ca2f78c3f3641410f19653de60 M	tests
```

The behavior of my program is as follows:
1. Create unix socket for internal communication
2. Create ConnectionPool with MySQLdb database driver. Please note that the driver is the standard one from MySQL-python, not the patched one from eventlet.
3. Main thread creates a GreenPool and spawns new greenthread for network probe (Netprobe thread)
4. Main thread listens on unix socket from step 1 for commands
5. Netprobe thread gets connection from ConnectionPool, runs a SELECT, and puts connection back in ConnectionPool.
6. Netprobe thread creates a eventlet.green.subprocess.Popen on /usr/bin/nmap . This command runs for 10 to 20 seconds.
7. About 5 seconds after subprocess, timeout exception occurs on failing versions of eventlet.
8. Based on nmap output, the netprobe thread spawns one or more greenthreads to probe specific IPs (IP probe thread).
9. The IP probe thread attempts to get a connection from ConnectionPool. On failing eventlet versions, this hangs.

**Comment by temoto** [comment]
Hello. Thank you for excellent bug report.

Overall description seems to fit known bug #50. Just to gather more information, please try again with environment `EVENTLET_HUB=poll` (since it worked twice in previous configurations).
Switching to green mysql-python or `eventlet.monkey_patch('MySQLdb')` ought to fix the symptoms you observe.

**Comment by avillacis** [comment]
I have tried with 0.16.1 . The EVENTLET_HUB environment variable has no effect. The "poll" and "selects" hubs hang in exactly the same way. I think this is a different bug from #50.

**Comment by avillacis** [comment]
Reverting the changes made by f37a87b1f8e2f09b8a30c2e947486cf0858650cd on tpool.py fixes the bug for me.

**Comment by avillacis** [comment]
This still happens with 0.18.4. Now the trace is slightly different:

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 457, in fire_timers
    timer()
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/timer.py", line 58, in __call__
    cb(*args, **kw)
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 154, in _timeout
    current.throw(exc)
  File "/usr/lib/python2.7/site-packages/eventlet/tpool.py", line 57, in tpool_trampoline
    _c = _rsock.recv(1)
  File "/usr/lib/python2.7/site-packages/eventlet/greenio/base.py", line 349, in recv
    return self._recv_loop(self.fd.recv, bufsize, flags)
  File "/usr/lib/python2.7/site-packages/eventlet/greenio/base.py", line 343, in _recv_loop
    self._read_trampoline()
  File "/usr/lib/python2.7/site-packages/eventlet/greenio/base.py", line 314, in _read_trampoline
    timeout_exc=socket.timeout("timed out"))
  File "/usr/lib/python2.7/site-packages/eventlet/greenio/base.py", line 203, in _trampoline
    mark_as_closed=self._mark_as_closed)
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 162, in trampoline
    return hub.switch()
  File "/usr/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 294, in switch
    return self.greenlet.switch()
timeout: timed out

```

**Comment by avillacis** [comment]
As before, EVENTLET_HUB has no effect in whether the exception occurs.

**Comment by avillacis** [comment]
I have now a minimal reproducer. This reproducer **requires** a MySQL database connection with valid credentials, as this is part of what triggers the bug.

This program is supposed to print "main: doing nothing..." and "do_something: doing nothing..." endlessly, until Ctrl-C is pressed. However, with the bug, a "timeout" traceback appears after 3 seconds, the value used in socket.settimeout() . Tested on CentOS 5 (runs correctly) and CentOS 7 (fails with 0.18.4).

```
#!/usr/bin/python
# vim: set fileencoding=utf-8 :

import sys
import signal
import eventlet
from eventlet.green import socket, select
from eventlet.greenpool import GreenPool
from eventlet.db_pool import ConnectionPool
import MySQLdb

cancelprobe = False
gpool = GreenPool()

def main():
	global cancelprobe

	socket.setdefaulttimeout(3)
	
	dbpool = ConnectionPool(MySQLdb,
				host = 'localhost',
				user = 'validuser',
				passwd = 'validpassword',
				db = 'somedatatabase',
                charset = 'utf8',
                use_unicode = True)
	
	gpool.spawn_n(do_something, dbpool)
	
	while gpool.running() > 0:
		try:
			sys.stdout.write('main: doing nothing...\n')
			lists = select.select([], [], [], 1)
		except (SystemExit, KeyboardInterrupt):
			sys.stderr.write('SystemExit/KeyboardInterrupt\n')
			cancelprobe = True

def do_something(dbpool):
	global cancelprobe
	
	dbconn = None
	try:
		dbconn = dbpool.get()
		dbpool.put(dbconn)
	except MySQLdb.Error, e:
		sys.stderr.write(str(e) + '\n')
		if dbconn != None: dbpool.put(dbconn)
		return
	dbconn = None
	
	while not cancelprobe:
		sys.stdout.write('do_something: doing nothing...\n')
		rd,wr,er = select.select([], [], [], 1)

if __name__ == '__main__':
    main()

```

**Comment by avillacis** [comment]
My test results:

I think you meant calling eventlet.monkey_patch() without arguments. The interpreter refuses an argument to eventlet.monkey_patch():
```
Traceback (most recent call last):
  File "./detect_endpoints_test", line 11, in <module>
    MySQLdb = eventlet.monkey_patch('MySQLdb')
TypeError: monkey_patch() takes exactly 0 arguments (1 given)
```

If I replace "import MySQLdb" with "MySQLdb = eventlet.import_patched('daemon')", I get this instead:
```
main: doing nothing...
/usr/lib64/python2.7/site-packages/MySQLdb/__init__.py:80: RuntimeWarning: Parent module 'MySQLdb' not found while handling absolute import
  from connections import Connection
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/eventlet/greenpool.py", line 82, in _spawn_n_impl
    func(*args, **kwargs)
  File "./detect_endpoints_test", line 44, in do_something
    dbconn = dbpool.get()
  File "/usr/lib/python2.7/site-packages/eventlet/db_pool.py", line 176, in get
    conn = super(BaseConnectionPool, self).get()
  File "/usr/lib/python2.7/site-packages/eventlet/pools.py", line 88, in get
    created = self.create()
  File "/usr/lib/python2.7/site-packages/eventlet/db_pool.py", line 276, in create
    self._db_module, self.connect_timeout, *self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/eventlet/db_pool.py", line 283, in connect
    conn = tpool.execute(db_module.connect, *args, **kw)
  File "/usr/lib/python2.7/site-packages/eventlet/tpool.py", line 125, in execute
    six.reraise(c, e, tb)
  File "/usr/lib/python2.7/site-packages/eventlet/tpool.py", line 83, in tworker
    rv = meth(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/MySQLdb/__init__.py", line 80, in Connect
    from connections import Connection
ImportError: No module named connections
```

Calling eventlet.monkey_patch() before the import MySQLdb does nothing. The program runs but still receives the erroneous timeout.

**Comment by temoto** [comment]
Your fix is merged in master cb8ac9b
Thank you.
